lib/api/model/submessage.dart

```diff
@@ -0,0 +1,285 @@
+import 'dart:convert';
+
+import 'package:json_annotation/json_annotation.dart';
+
+part 'submessage.g.dart';
+
+/// Data used for certain experimental Zulip widgets including polls and todo
+/// lists.
+///
+/// See:
+///   https://zulip.com/api/get-messages#response
+///   https://zulip.readthedocs.io/en/stable/subsystems/widgets.html
+@JsonSerializable(fieldRename: FieldRename.snake)
+class Submessage {
+  const Submessage({
+    required this.msgType,
+    required this.content,
+    required this.messageId,
+    required this.senderId,
+    required this.id,
+  });
+
+  @JsonKey(unknownEnumValue: SubmessageType.unknown)
+  final SubmessageType msgType;
+  /// [SubmessageData] encoded in JSON.
+  @JsonKey(readValue: readContent, toJson: contentToJson)
+  final Object? content;
```

Comment:
Storing the parsed data back to `content` might be redundant, because these `Submessage` objects that are meant to be intermediate. I guess an alternative would be just storing the JSON string, and parse them when producing the `SubmessageData`.

Reply:
> Storing the parsed data back to `content` might be redundant, because these `Submessage` objects that are meant to be intermediate.

Hmm, I don't really follow this. What would go wrong in putting the fully-parsed `SubmessageData` object here?

Reply:
Wasn't clear, sorry. We can't parse `SubmessageData` without looking at the content of the other submessages on the message. To parse it, we need to look at the index of the submessage, and the type of the Zulip widget (indicated by the content of the first submessage). These are something that can't be handled by `Submessage.fromJson` alone.

This kind of context aware parsing requirement makes it more appropriate to parse the `SubmessageData` on a method implemented by the `Message` class. So it is more of a restriction that we can't put fully-parsed `SubmessageData` there.

Reply:
I see, annoying. (I think I've mentioned previously that this "submessages" API is not our best work.)

Let's have a comment explaining why `fromJson` can't fully decode this for us — the fact that it needs that context from outside. In particular I think the case where that constraint really bites is a SubmessageEvent, where we don't even know whether it's a poll widget or a todo widget until we look up the message ID in the message store.

And then yeah, I agree it may be cleaner to store just the JSON string here and let the consumer call `jsonDecode` as well as whatever "fromJson" sort of method it wants to call on the result of that. But if switching to that way seems like a pain, this way is OK too.

Reply:
Updated the PR to have a comment explaining that.